include/shark/Algorithms/DirectSearch/Operators/Hypervolume/HypervolumeSubsetSelection2D.h

```diff
@@ -0,0 +1,298 @@
+	
+		double a;
+		double b;
+		std::size_t index;
+		
+		LinearFunction(double a, double b, std::size_t index = 0):a(a), b(b), index(index){}
+		LinearFunction(){}
+		
+		double eval(double x)const{
+			return a*x + b;
+		}
+	};
+
+	/// \brief Returns the intersection of two linear functions
+	double Intersection(LinearFunction f1, LinearFunction f2)const{
+		return (f2.b - f1.b) / (f1.a - f2.a);
+	}
+	
+	
+	/// \brief  Calculates for each given x the maximum among the functions f, i.e. the upper envelope of f.
+	/// 
+	/// Algoruthm 2 in the paper. Complexity O(n)
+	/// given a set of functions f_1...f_n, ordered by slope such that f_1.a < f_2.a<... and points with x-coordinate x_1<...<x_n
+	/// computes h_i = max_{1 <= j < i} f_j(x_i) for i=1,...,n as well as the index of the function leading to the value h_i
+	std::pair<std::vector<double>,std::vector<std::size_t> > upperEnvelope(
+		std::vector<LinearFunction>const& functions,
+		std::vector<Point> const& points
+	)const{
+		std::size_t n = points.size();
+		std::vector<double> h(n);
+		std::vector<std::size_t> chosen(n);
+		std::deque<LinearFunction> s;
+
+		// This is the original algorithm 2 as in the paper. Even if the paper looks at maximum
+		// hypervolume where domination is given when one point has LARGER function
+		// values as the other, In section 3.2 they transform the problem to a problem
+		// where domination is given by SMALLER function values and then accordingly
+		// give the algorithm for this type. They just give the transformation but do not say
+		// what the transformation does so it is not clear until you implement it.
+		//
+		// This is a super confusing part of the paper, please kids, do not be like Bringmann et al.
+		// Keep it simple, stupid. Sometimes an additional index does not hurt.
+		//
+		//the algorithm works by inserting functions f_1 to f_i one-by-one, figuring out which functions
+		// are dominated (not being part of the upper envelope) and removing all functions which for
+		// function values x_i,x_i+1,... are already smaller than one of the other function.
+		// using the ordering relations given the set s contains the function ordered by (current) function value.
+		// so after iteration i we can just extract the largest function value for x_i by looking at the first element of s.
+		for (std::size_t i = 0; i != n; ++i) {
+
+			// remove dominated functions.
+			// as we push back into s,
+			// at the end of s are the functions with largest slope.
+			// therefore if we have the last two elements as s_-1 and s_-2 and the new
+			// function f, knowing that the intersection of s_-1 and f is smaller than the intersection
+			// of s_-1 and s_-2 means that s_-1 is dominated  by s_-2 and f and thus can be removed.
+			while (s.size() > 1 ) {
+				
+				double d1 = Intersection(functions[i], s.end()[-1]);
+				double d2 = Intersection(s.end()[-2], s.end()[-1]);
+
+				if (d1 <= d2 || std::abs(d1-d2) < 1.e-10) {//check for numeric stability
```

Comment:
This can be simplified to the condition d1 <= d2 + 1e-10, but we can also leave this optimization to the compiler because the current code make the intent more clear.